### PR TITLE
AuthorityKeyIdentifier support

### PR DESCRIPTION
**Pull request by reaperhulk**
refs #1743

### CODE REVIEW

File: docs/x509.rst [hunk 1]
```diff
@@ -577,6 +577,25 @@ X.509 Extensions
...
+
+        :type: bytes
+
+    .. attribute:: authority_cert_issuer
+
+        :type: :class:`Name` or None
+
+    .. attribute:: authority_cert_serial_number
+
+        :type: int or None
```
Comment (alex) [inline docs/x509.rst]: Need some prose describing what each of these things are.

**Comment by alex** [comment]
Besides the prose, I think this is ok.

File: docs/x509.rst [hunk 2]
```diff
@@ -652,6 +652,32 @@ X.509 Extensions
...
+
+    The authority key identifier extension provides a means of identifying the
+    public key corresponding to the private key used to sign a certificate.
+
+    .. attribute:: key_identifier
+
+        :type: bytes
+
+        A value derived from the public key used to verify the certificate's
+        signature.
```
Comment (alex) [inline docs/x509.rst]: I'm not sure I quite understand this, what is someone supposed to do with this value?

**Comment by alex** [comment]
retest this please

**Comment by alex** [comment]
retest this please

**Comment by alex** [comment]
retest this please

**Comment by reaperhulk** [comment]
retest this please

**Comment by reaperhulk** [comment]
retest this please

File: docs/x509.rst (hunk 2 above)
Comment (reaperhulk) [inline docs/x509.rst]: It's meant to assist in chain building when the normal methods (look up the issuer) may result in multiple potential chain paths.

Comment (alex) [inline docs/x509.rst]: Can you explain what "A value derived from" is supposed to mean?

On Sat, Apr 18, 2015 at 9:49 PM, Paul Kehrer <notifications@github.com>
wrote:

> In docs/x509.rst
> <https://github.com/pyca/cryptography/pull/1835#discussion_r28649039>:
>
> > @@ -652,6 +652,32 @@ X.509 Extensions
> >      purposes indicated in the key usage extension. The object is
> >      iterable to obtain the list of :ref:`extended key usage OIDs <eku_oids>`.
> >
> > +.. class:: AuthorityKeyIdentifier
> > +
> > +    .. versionadded:: 0.9
> > +
> > +    The authority key identifier extension provides a means of identifying the
> > +    public key corresponding to the private key used to sign a certificate.
> > +
> > +    .. attribute:: key_identifier
> > +
> > +        :type: bytes
> > +
> > +        A value derived from the public key used to verify the certificate's
> > +        signature.
>
> It's meant to assist in chain building when the normal methods (look up
> the issuer) may result in multiple potential chain paths.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1835/files#r28649039>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

Comment (reaperhulk) [inline docs/x509.rst]: The spec is noncommittal on what that means, unfortunately. There's a "standard" way that is pretty much what everybody does (to my knowledge) but technically you can put whatever the hell you want here. RFC 5280 is not great sometimes.

Comment (alex) [inline docs/x509.rst]: ... wtf RFC 5280. Can we get some explicit language about how the contents
of this is allowed to verify or something.

On Sat, Apr 18, 2015 at 10:06 PM, Paul Kehrer <notifications@github.com>
wrote:

> In docs/x509.rst
> <https://github.com/pyca/cryptography/pull/1835#discussion_r28649105>:
>
> > @@ -652,6 +652,32 @@ X.509 Extensions
> >      purposes indicated in the key usage extension. The object is
> >      iterable to obtain the list of :ref:`extended key usage OIDs <eku_oids>`.
> >
> > +.. class:: AuthorityKeyIdentifier
> > +
> > +    .. versionadded:: 0.9
> > +
> > +    The authority key identifier extension provides a means of identifying the
> > +    public key corresponding to the private key used to sign a certificate.
> > +
> > +    .. attribute:: key_identifier
> > +
> > +        :type: bytes
> > +
> > +        A value derived from the public key used to verify the certificate's
> > +        signature.
>
> The spec is noncommittal on what that means, unfortunately. There's a
> "standard" way that is pretty much what everybody does (to my knowledge)
> but technically you can put whatever the hell you want here. RFC 5280 is
> not great sometimes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1835/files#r28649105>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

Comment (alex) [inline docs/x509.rst]: And maybe what the standard thing people do is? (Is there a way to know
that it's what they're doing?)

On Sat, Apr 18, 2015 at 10:08 PM, Alex Gaynor <alex.gaynor@gmail.com> wrote:

> ... wtf RFC 5280. Can we get some explicit language about how the contents
> of this is allowed to verify or something.
>
> On Sat, Apr 18, 2015 at 10:06 PM, Paul Kehrer <notifications@github.com>
> wrote:
>
>> In docs/x509.rst
>> <https://github.com/pyca/cryptography/pull/1835#discussion_r28649105>:
>>
>> > @@ -652,6 +652,32 @@ X.509 Extensions
>> >      purposes indicated in the key usage extension. The object is
>> >      iterable to obtain the list of :ref:`extended key usage OIDs <eku_oids>`.
>> >
>> > +.. class:: AuthorityKeyIdentifier
>> > +
>> > +    .. versionadded:: 0.9
>> > +
>> > +    The authority key identifier extension provides a means of identifying the
>> > +    public key corresponding to the private key used to sign a certificate.
>> > +
>> > +    .. attribute:: key_identifier
>> > +
>> > +        :type: bytes
>> > +
>> > +        A value derived from the public key used to verify the certificate's
>> > +        signature.
>>
>> The spec is noncommittal on what that means, unfortunately. There's a
>> "standard" way that is pretty much what everybody does (to my knowledge)
>> but technically you can put whatever the hell you want here. RFC 5280 is
>> not great sometimes.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/pyca/cryptography/pull/1835/files#r28649105>.
>>
>
>
>
> --
> "I disapprove of what you say, but I will defend to the death your right
> to say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
> "The people's good is the highest law." -- Cicero
> GPG Key fingerprint: 125F 5C67 DFE9 4084
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

Comment (reaperhulk) [inline docs/x509.rst]: Found a section in the RFC that talks about how it should be generated, so linked that.

File: docs/x509.rst [hunk 3]
```diff
@@ -652,6 +652,33 @@ X.509 Extensions
...
+    The authority key identifier extension provides a means of identifying the
+    public key corresponding to the private key used to sign a certificate.
+
+    .. attribute:: key_identifier
+
+        :type: bytes
+
+        A value derived from the public key used to verify the certificate's
+        signature. See :rfc:`5280` section 4.2.1.2 for more details on the
+        recommended way to compute this value.
```
Comment (alex) [inline docs/x509.rst]: Sorry, but I'm still not really liking this paragraph. It doesn't explain to me what I do if I just parsed a certificate with some value here

Comment (reaperhulk) [inline docs/x509.rst]: I think explaining the semantics of how a certificate chain may be built is way too complex to fit in this section. We have links to normative references where things (like this) can't be simplified. "What should I do if I parsed a certificate with some value" here is basically "you shouldn't touch this, it's not for you".

Maybe simply stating "this extension is typically used by software to assist in determining the appropriate certificate chain" would be sufficient?

Comment (alex) [inline docs/x509.rst]: That, plus a link to the place in the RFC that explains it should be ok.

On Sun, Apr 19, 2015 at 9:55 AM, Paul Kehrer <notifications@github.com>
wrote:

> In docs/x509.rst
> <https://github.com/pyca/cryptography/pull/1835#discussion_r28653010>:
>
> > @@ -652,6 +652,33 @@ X.509 Extensions
> >      purposes indicated in the key usage extension. The object is
> >      iterable to obtain the list of :ref:`extended key usage OIDs <eku_oids>`.
> >
> > +.. class:: AuthorityKeyIdentifier
> > +
> > +    .. versionadded:: 0.9
> > +
> > +    The authority key identifier extension provides a means of identifying the
> > +    public key corresponding to the private key used to sign a certificate.
> > +
> > +    .. attribute:: key_identifier
> > +
> > +        :type: bytes
> > +
> > +        A value derived from the public key used to verify the certificate's
> > +        signature. See :rfc:`5280` section 4.2.1.2 for more details on the
> > +        recommended way to compute this value.
>
> I think explaining the semantics of how a certificate chain may be built
> is way too complex to fit in this section. We have links to normative
> references where things (like this) can't be simplified. "What should I do
> if I parsed a certificate with some value" here is basically "you shouldn't
> touch this, it's not for you".
>
> Maybe simply stating "this extension is typically used by software to
> assist in determining the appropriate certificate chain" would be
> sufficient?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1835/files#r28653010>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

File: docs/x509.rst [hunk 4]
```diff
@@ -938,3 +967,4 @@ Exceptions
 
 .. _`public key infrastructure`: https://en.wikipedia.org/wiki/Public_key_infrastructure
 .. _`TLS`: https://en.wikipedia.org/wiki/Transport_Layer_Security
+.. _`RFC 5280 section 4.2.1.1`: http://tools.ietf.org/html/rfc5280#section-4.2.1.1
```
Comment (alex) [inline docs/x509.rst]: https tho

**Pull request merged**